### PR TITLE
[workingset] purge actpool when run action fails

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
when `runAction()` fails, the tx got stuck, and prevent block from being produced. So the erroneous tx (and other pending tx from same sender) should be removed from actpool

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: state/factory/workingset.go [hunk 1]
```diff
@@ -591,22 +599,41 @@ func (ws *workingSet) pickAndRunActions(
...
+					}
+					actionIterator.PopAccount()
+					nextActionHash, hashErr := nextAction.Hash()
+					if hashErr != nil {
+						return nil, errors.Wrapf(hashErr, "Failed to get hash for %x", nextActionHash)
+					}
+					return nil, errors.Wrapf(err, "Failed to update state changes for selp %x", nextActionHash)
+				}
+			} else {
+				switch errors.Cause(err) {
```
Comment (dustinxie) [inline state/factory/workingset.go:622]: else part is current code

File: state/factory/workingset.go [hunk 2]
```diff
@@ -591,22 +599,41 @@ func (ws *workingSet) pickAndRunActions(
...
-				continue
-			default:
-				nextActionHash, hashErr := nextAction.Hash()
-				if hashErr != nil {
-					return nil, errors.Wrapf(hashErr, "Failed to get hash for %x", nextActionHash)
+			if fCtx.PurgeActpoolOnRunActionError {
+				switch errors.Cause(err) {
+				case nil:
+					// do nothing
+				case action.ErrChainID, action.ErrGasLimit, errUnfoldTxContainer, errDeployerNotWhitelisted:
```
Comment (dustinxie) [inline state/factory/workingset.go:607]: should we also call `ap.DeleteAction()` here?

File: state/factory/workingset.go [hunk 3]
```diff
@@ -591,22 +599,41 @@ func (ws *workingSet) pickAndRunActions(
...
-			case action.ErrChainID:
-				continue
-			case action.ErrGasLimit, errUnfoldTxContainer:
-				actionIterator.PopAccount()
-				continue
-			default:
-				nextActionHash, hashErr := nextAction.Hash()
-				if hashErr != nil {
-					return nil, errors.Wrapf(hashErr, "Failed to get hash for %x", nextActionHash)
+			if fCtx.PurgeActpoolOnRunActionError {
```
Comment (envestcc) [inline state/factory/workingset.go:603]: may not be a HF

File: state/factory/workingset.go (hunk 2 above)
Comment (envestcc) [reply state/factory/workingset.go:607]: should remove from actpool

Comment (envestcc) [reply state/factory/workingset.go:607]: but the ErrGasLimit action may run successfully in next block?

File: state/factory/workingset.go [hunk 4]
```diff
@@ -591,22 +599,41 @@ func (ws *workingSet) pickAndRunActions(
...
+				case nil:
+					// do nothing
+				case action.ErrChainID, action.ErrGasLimit, errUnfoldTxContainer, errDeployerNotWhitelisted:
+					actionIterator.PopAccount()
+					continue
+				default:
+					if caller := nextAction.SenderAddress(); caller != nil {
+						ap.DeleteAction(caller)
+					}
+					actionIterator.PopAccount()
```
Comment (CoderZhi) [inline state/factory/workingset.go:614]: if an error will be returned eventually, no need to `PopAccount`

File: state/factory/workingset.go (hunk 2 above)
Comment (CoderZhi) [reply state/factory/workingset.go:607]: not for `action.ErrGasLimit`

**Review by CoderZhi — COMMENTED** [review]
I don't think we need a hard fork tag

File: state/factory/workingset.go [hunk 5]
```diff
@@ -169,7 +170,11 @@ func (ws *workingSet) runAction(
 	// for replay tx, check against deployer whitelist
 	g := genesis.MustExtractGenesisContext(ctx)
 	if selp.Encoding() == uint32(iotextypes.Encoding_ETHEREUM_UNPROTECTED) && !g.IsDeployerWhitelisted(selp.SenderAddress()) {
-		return nil, errors.Errorf("replay deployer %v not whitelisted", selp.SenderAddress().String())
+		if protocol.MustGetFeatureCtx(ctx).PurgeActpoolOnRunActionError {
+			return nil, errors.Wrap(errDeployerNotWhitelisted, selp.SenderAddress().String())
+		} else {
+			return nil, errors.Errorf("replay deployer %v not whitelisted", selp.SenderAddress().String())
```
Comment (CoderZhi) [inline state/factory/workingset.go:176]: it is not a hard fork. If an error is returned during block mint, the action will not be included. If an error is returned during block validation, the block will be rejected

File: action/protocol/context.go [hunk 6]
```diff
@@ -127,6 +127,7 @@ type (
 		MigrateNativeStake                      bool
 		AddClaimRewardAddress                   bool
 		EnforceLegacyEndorsement                bool
+		PurgeActpoolOnRunActionError            bool
```
Comment (CoderZhi) [inline action/protocol/context.go:130]: thus, we can delete the tag

**Review by CoderZhi — COMMENTED** [review]
to keep record, it is better not to overwrite commits with comments.

File: state/factory/workingset.go (hunk 5 above)
Comment (dustinxie) [reply state/factory/workingset.go:176]: yeah, fixed

File: action/protocol/context.go (hunk 6 above)
Comment (dustinxie) [reply action/protocol/context.go:130]: yeah, it does not need to be a hard-fork

**Review by CoderZhi — APPROVED** [review]

File: state/factory/workingset.go [hunk 7]
```diff
@@ -591,17 +595,24 @@ func (ws *workingSet) pickAndRunActions(
...
 				continue
 			}
-			// TODO: error may cause the tx to be stuck in the action pool until expired
 			receipt, err := ws.runAction(actionCtx, nextAction)
 			switch errors.Cause(err) {
 			case nil:
 				// do nothing
-			case action.ErrChainID:
+			case action.ErrGasLimit:
+				actionIterator.PopAccount()
```
Comment (envestcc) [inline state/factory/workingset.go:603]: In an extreme scenario, if this is a huge action that exceeds the entire block gas limit, then this transaction will remain in the actpool, blocking this user's transaction indefinitely.

**Review by envestcc — APPROVED** [review]

**Pull request merged**